e2e/extensions-protocol.e2e.spec.ts

```diff
@@ -0,0 +1,118 @@
+  createDefaultSettings, getResourceBinDir, retry, startRancherDesktop, teardown,
+} from './utils/TestUtils';
+
+import { ContainerEngine, Settings } from '@pkg/config/settings';
+import { spawnFile } from '@pkg/utils/childProcess';
+
+/** The top level source directory, assuming we're always from the tree */
+const srcDir = path.dirname(path.dirname(__filename));
+const rdctl = executable('rdctl');
+
+/**
+ * Get the given executable. Similar to @pkg/utils/resources, but does not use
+ * Electron.app (which doesn't work during the test).
+ */
+function executable(name: string) {
+  const exeName = name + (process.platform === 'win32' ? '.exe' : '');
+
+  return path.join(srcDir, 'resources', process.platform, 'bin', exeName);
+}
+
+test.describe.serial('Extensions protocol handler', () => {
+  let app: ElectronApplication;
+  let page: Page;
+  let isContainerd = false;
+
+  async function ctrctl(...args: string[]) {
+    let tool = executable('nerdctl');
+
+    if (!isContainerd) {
+      tool = executable('docker');
+      if (process.platform !== 'win32') {
+        args = ['--context', 'rancher-desktop'].concat(args);
+      }
+    }
+
+    return await spawnFile(tool, args,
+      {
+        stdio: 'pipe',
+        env:   {
+          ...process.env,
+          PATH: `${ process.env.PATH }${ path.delimiter }${ getResourceBinDir() }`,
+        },
+      });
+  }
+
+  test.beforeAll(async() => {
+    createDefaultSettings({ kubernetes: { enabled: false } });
+    app = await startRancherDesktop(__filename, { mock: false });
+    page = await app.firstWindow();
+  });
+
+  test.afterAll(() => teardown(app, __filename));
+
+  test('should load backend', async() => {
+    await (new NavPage(page)).progressBecomesReady();
+  });
+
+  test('determine container engine in use', async() => {
+    const { stdout } = await spawnFile(rdctl, ['list-settings'], { stdio: 'pipe' });
+    const settings: Settings = JSON.parse(stdout);
+
+    expect(settings.containerEngine.name).toMatch(/^(?:containerd|moby)$/);
+    isContainerd = settings.containerEngine.name === ContainerEngine.CONTAINERD;
+  });
+
+  test('wait for buildkit', async() => {
+    test.skip(!isContainerd, 'Not running containerd, no need to wait for buildkit');
+
+    await retry(() => spawnFile(rdctl, ['shell', 'buildctl', 'debug', 'info']));
+  });
+
+  test('wait for docker context', async() => {
+    test.skip(isContainerd, 'Not running moby, no need to wait for context');
+    test.skip(process.platform === 'win32', 'Not setting context on Windows');
+
+    await retry(() => ctrctl('context', 'inspect', 'rancher-desktop'));
+  });
+
+  test('wait for docker daemon to be up', async() => {
+    test.skip(isContainerd, 'Not running moby, no need to wait for context');
+
+    // On Windows, the docker proxy can flap for a while. So we try a few times
+    // in a row (with pauses in the middle) to ensure the backend is stable
+    // before continuing.
+    for (let i = 0; i < 10; ++i) {
```

Comment:
Why not put the cycles and timeouts in `retry` directly? Wouldn't
```
retry(() => ctrctl('system', 'info'), { delay: 500: tries: 300 })
```
more or less accomplish the same thing?

Reply:
No, the point is that it's _flapping_, so succeeding once doesn't mean it's going to work. I need to wait until it succeeds multiple times before continuing.

Reply:
ok that makes sense